runtime/startup_aux.c

```diff
@@ -93,10 +95,12 @@ void caml_parse_ocamlrunparam(void)
       switch (*opt++){
       case 'b': scanmult (opt, &params.backtrace_enabled); break;
       case 'c': scanmult (opt, &params.cleanup_on_exit); break;
+      case 'd': scanmult (opt, &params.max_domains); break;
       case 'l': scanmult (opt, &params.init_max_stack_wsz); break;
       case 'M': scanmult (opt, &params.init_custom_major_ratio); break;
       case 'm': scanmult (opt, &params.init_custom_minor_ratio); break;
       case 'n': scanmult (opt, &params.init_custom_minor_max_bsz); break;
+      case 'N': scanmult (opt, &params.minor_heap_max_wsz); break;
```

Comment:
Isn't there an interaction between `N` and `s` parameters? For soundness, `s <= N` must hold. I see three options here:

1. The runtime can raise an exception if this doesn't hold. Hard for the user to get common cases right. 
2. Silently bump `N` to be at least as much as `s`. Choose the default values such as `N >= s`. 
3. Remove the `N` parameter altogether and only have `s`. Interpret `N == s`. The downside here is that the size of the minor heap cannot be programmatically increased during runtime from its initial value since the minor heap is already at the maximum size. It can only be decreased. FWIW, depending on the default value for `N` in option 2, it will also suffer from the same problem.

If it is not common to increase the size of the minor heap during execution, I would go for 3 as it has one fewer option, and is less confusing for the users.

Reply:
@damiendoligez might have an opinion on this.

Reply:
Searching the Internet, I can't seem to find any discussions where people suggest increasing the minor heap size during execution. I only saw people recommending to increase the minor heap size using the `s` option.

Are there other ways we could go about checking? Is it possible to search all opam packages for code that increases the size of the minor heap during execution?

Clearly, 3 has the least potential for confusing users.

Reply:
I have a program that has command-line parameters to tune the gc control knobs (see https://github.com/Gbury/dolmen/blob/7cd59bbeaa536056d19facdbdc724b104e891732/src/bin/options.ml#L252-L262 ,
https://github.com/Gbury/dolmen/blob/7cd59bbeaa536056d19facdbdc724b104e891732/src/bin/options.ml#L316 , and https://github.com/Gbury/dolmen/blob/7cd59bbeaa536056d19facdbdc724b104e891732/src/bin/options.ml#L369-L401 ), but i'ts mainly because I find it easier to use command line arguments rather than env variables, and because once upon a time, I ran some benchmarks to tweak the default parameters a bit. On that note, there's also at least Coq that increases the minor heap size at startup (but I think it falls in the definition of during execution), see https://github.com/coq/coq/blob/07537426ad106c721e67d513c6db2e9047131e7b/sysinit/coqinit.ml

Reply:
CompCert also increases minor heap size at start-up:
https://github.com/AbsInt/CompCert/blob/85b1c4091e13dec13fe03f28e81b256c60f9f7ef/driver/Driver.ml#L403-L406

You could OPAM-grep for `Gc.set`: I suspect many uses of this function tweak the minor heap size.

Reply:
I'm running the OPAM-grep, and so far, it looks like you are correct:

Using `Gc.set` to set the minor heap size is something that people commonly do.

Reply:
Changing heap sizes throughout runtime has been proposed/discussed too: https://sympa.inria.fr/sympa/arc/caml-list/2010-11/msg00323.html. I have sometimes had success with similar techniques, but would not suggest them as a general-purpose default.